### PR TITLE
Feat/simple mse metric #388

### LINKED ISSUE #388: [FEATURE] Add Davidsonian Scene Graph (DSG) Metric
‼️ If you want to work on this issue: please comment below and wait until a maintainer assigns this issue to you before opening a PR to avoid several contributions on the same issue. Thanks! 😊  

---

## ✨ **What You’ll Do**  
Right now, Pruna has several evaluation metrics, but it lacks a semantic dependency-graph based metric. The **Davidsonian Scene Graph (DSG)** metric ([paper](https://arxiv.org/abs/2310.18235)) builds a dependency graph of atomic, non-redundant questions to check semantic consistency in generated images.  

Your task will be to:  
- Add a new metric class `DavidsonianSceneGraphMetric` under `pruna/metrics/`.  
- Implement the standard metric interface (`update(...)`, `compute(...)`).  
- In `update(...)`, for each prompt + image, generate atomic questions, run VQA, and build/score the dependency graph as described in the paper.  
- Integrate this metric into Pruna’s registry so it can be selected in evaluation runs.  
- Add tests that validate correctness on toy examples and edge cases.  
- Document the new metric and provide a simple usage example.  

---

## 🤖 **Useful Resources**  
- [Davidsonian Scene Graph paper](https://arxiv.org/abs/2310.18235) for methodology and examples.  
- Existing metrics in `pruna/metrics` as references for structure.  
- Question Generation (QG) and Visual Question Answering (VQA) models/tools to implement the question-answer loop.  
- Pruna’s metric registry and how new metrics are registered.  

---

## ✅ **Acceptance Criteria**  
- It follows the style guidelines.  
- The new `DavidsonianSceneGraphMetric` is documented in code and user docs.  
- Tests demonstrate that the metric computes expected scores on simple controlled examples.  
- The metric integrates properly into Pruna’s evaluation framework.  

And don’t forget to give us a ⭐️!
---

## **❓ Questions?**

Feel free to jump into the [#contributing Discord channel](https://discord.com/invite/JFQmtFKCjd) if you hit any roadblocks. Can’t wait to see your contribution! 🚀

---

## Share on Socials

[Share on X](https://twitter.com/intent/tweet?text=I%20just%20contributed%20to%20%40PrunaAI%20open%20source%20project!%20%F0%9F%9A%80%20Check%20out%20my%20PR%20%F0%9F%91%89%20%5BADD%20LINK%20TO%20PR%5D)

### PR DESCRIPTION
**Pull request by AnikethBhosale**
## Description
Implements MSE (Mean Squared Error) metric for Pruna's evaluation framework. The metric computes mean squared error between model predictions and ground truth values, accumulating results across batches using StatefulMetric pattern.

## Related Issue
Fixes #388

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
- Created comprehensive test suite with 15 tests covering:
  - Perfect match scenarios (MSE = 0)
  - Known value calculations
  - Multiple batch accumulation
  - Empty state handling
  - Multi-dimensional tensors (1D to 4D)
  - Device compatibility (CPU/CUDA)
  - Edge cases (None inputs, shape mismatches)
- All tests pass: 14 passed, 1 skipped (CUDA test on non-CUDA systems)
- Test coverage: 89% for metric_mse.py
- Verified style compliance: ty check and ruff check pass

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Additional Notes
- Metric is registered with MetricRegistry and can be used via `Task(metrics=["mse"])`
- Follows existing patterns from other stateful metrics (e.g., SharpnessMetric)
- Uses list-based state accumulation as required by StatefulMetric framework
- Documentation includes usage examples, technical details, and related metrics

----
# Note : Even if this is not merged, do use the screenshot attached in the CONTRIBUTING.md (at the end) which was added as requested by @johannaSommer on pruna's discord server (resources channel)

### CODE REVIEW

File: src/pruna/evaluation/metrics/metric_mse.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        inputs = metric_data_processor(x, gt, outputs, self.call_type)
+        gt_tensor = inputs[0]
+        output_tensor = inputs[1]
+
+        if gt_tensor is None or output_tensor is None:
+            pruna_logger.debug("MSE.update received None for gt or outputs; skipping.")
+            return
+
+        # Ensure tensors are on the same device
+        output_tensor = output_tensor.to(gt_tensor.device)
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_mse.py]: this is a great idea, that's why we have integrated device casting in the metric_data_processor. how do you feel about passing the device to it instead?

File: src/pruna/evaluation/metrics/metric_mse.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        ----------
+        x : Any | Tensor
+            The input data (may be unused depending on call_type).
+        gt : Tensor
+            The ground truth values.
+        outputs : Tensor
+            The model predictions/outputs.
+        """
+        # Process inputs based on call_type (returns tuple of tensors)
+        inputs = metric_data_processor(x, gt, outputs, self.call_type)
```
Comment (begumcig) [inline src/pruna/evaluation/metrics/metric_mse.py]: you can pass the device here (regarding the comment below)

File: tests/evaluation/test_mse.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
+# Copyright 2025 - Pruna AI GmbH. All rights reserved.
```
Comment (begumcig) [inline tests/evaluation/test_mse.py:1]: I really like the variety in the tests! How do you feel about testing with some data from the pruna similar to what we have in  `tests/evaluation/test_torch_metrics.py`?

File: MSE_IMPLEMENTATION_SUMMARY.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
+# MSE Metric Implementation Summary
```
Comment (begumcig) [inline MSE_IMPLEMENTATION_SUMMARY.md:1]: Thank you a lot for this detailed summary, are we planning on merging it to Pruna or is it more for giving information? I think this would be even more beneficial as the PR description

**Comment by AnikethBhosale** [comment]
@begumcig , I've been a bit busy because of some interviews, will look into this shortly .
As for your last question , I was hoping that this gets merged into pruna.

**Comment by davidberenstein1957** [comment]
@begumcig @AnikethBhosale any follow-ups needed to unblock this? :)

**Comment by AnikethBhosale** [comment]
@davidberenstein1957 I have made changes acccording to the previous review. I do not know how to unblock this.
